File: src/engine/ChangeLens.Engine/AnalysisRuns/Models/AnalysisStartResult.cs
using System.Text.Json.Serialization;

namespace ChangeLens.Engine.AnalysisRuns.Models;

/// <summary>
///     Represents the tagged analysis-start outcome union in the engine protocol.
/// </summary>
[JsonPolymorphic(TypeDiscriminatorPropertyName = "state")]
[JsonDerivedType(typeof(AcceptedAnalysisStartResult), "accepted")]
[JsonDerivedType(typeof(RejectedStaleAnalysisStartResult), "rejectedStale")]
[JsonDerivedType(typeof(RejectedActiveAnalysisStartResult), "rejectedActive")]
internal abstract record AnalysisStartResult;

File: src/desktop/src-tauri/src/analysis/models/analysis_start_result.rs
use crate::analysis::models::validation::deserialize_run_id;
use serde::{Deserialize, Serialize};

/// Represents the tagged analysis-start outcome union in the engine protocol.
#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Serialize)]
#[serde(
    tag = "state",
    rename_all = "camelCase",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum AnalysisStartResult {
    Accepted {
        #[serde(deserialize_with = "deserialize_run_id")]
        run_id: String,
        requested_at: u64,
    },
    RejectedStale,
    RejectedActive {
        #[serde(deserialize_with = "deserialize_run_id")]
        active_run_id: String,
    },
}
